interactions/client.py

```diff
@@ -722,25 +763,117 @@ def __init__(self, client):
-                else:
-                    cls.commands.append(content)
 
-        for _command in cls.commands:
-            cls.client.command(**_command)
+    def __new__(cls, client: Client, *args, **kwargs) -> "Extension":
+
+        self = super().__new__(cls)
+
+        self.client = client
+        self._commands = {}
+        self._listeners = {}
+        self._components = {}
+
+        # This gets every coroutine in a way that we can easily change them
+        # cls
+        for name, func in inspect.getmembers(self, predicate=iscoroutinefunction):
+
+            # TODO we can make these all share the same list, might make it easier to load/unload
+            if hasattr(func, "__listener_name__"):  # set by extension_listener
+                func = client.event(
+                    func, name=func.__listener_name__
+                )  # capture the return value for friendlier ext-ing
+
+                listeners = self._listeners.get(func.__listener_name__, [])
+                listeners.append(func)
+                self._listeners[func.__listener_name__] = listeners
+
+            if hasattr(func, "__command_data__"):  # Set by extension_command
+                args, kwargs = func.__command_data__
+                func = client.command(*args, **kwargs)(func)
+
+                cmd_name = f"command_{kwargs.get('name') or func.__name__}"
+
+                commands = self._commands.get(cmd_name, [])
+                commands.append(func)
+                self._commands[cmd_name] = commands
+
+            if hasattr(func, "__component_data__"):
+                args, kwargs = func.__component_data__
+                func = client.component(*args, **kwargs)(func)
+
+                component = kwargs.get("component") or args[0]
+                comp_name = (
+                    _component(component).custom_id
+                    if isinstance(component, (Button, SelectMenu))
+                    else component
+                )
+                comp_name = f"component_{comp_name}"
+
+                components = self._components.get(comp_name, [])
+                components.append(func)
+                self._components[comp_name] = components
+
+        client.extensions[cls.__name__] = self
+
+        return self
+
+    def teardown(self):
+        for event, funcs in self._listeners.items():
+            for func in funcs:
+                self.client.websocket.dispatch.events[event].remove(func)
+
+        for component, funcs in self._components.items():
+            for func in funcs:
+                self.client.websocket.dispatch.events[component].remove(func)
+
+        for cmd, funcs in self._commands.items():
+            for func in funcs:
+                self.client.websocket.dispatch.events[cmd].remove(func)
+
+        clean_cmd_names = ["_".join(cmd.split("_")[1:]) for cmd in self._commands.keys()]
+        cmds = filter(
+            lambda x: x["name"] in clean_cmd_names,
```

Comment:
```suggestion
            lambda cmd: cmd["name"] in clean_cmd_names,
```

Try and step away from alphabet soup abstraction, `x` may make sense from a maintainer's perspective but anyone else reading this can easily get confused by common iterator naming conventions